src/interface/sparse_pack_base.hpp

```diff
@@ -0,0 +1,199 @@
+// Alamos National Laboratory (LANL), which is operated by Triad National Security, LLC
+// for the U.S. Department of Energy/National Nuclear Security Administration. All rights
+// in the program are reserved by Triad National Security, LLC, and the U.S. Department
+// of Energy/National Nuclear Security Administration. The Government is granted for
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+#ifndef INTERFACE_SPARSE_PACK_BASE_HPP_
+#define INTERFACE_SPARSE_PACK_BASE_HPP_
+
+#include <algorithm>
+#include <functional>
+#include <limits>
+#include <map>
+#include <memory>
+#include <regex>
+#include <string>
+#include <tuple>
+#include <type_traits>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+#include "coordinates/coordinates.hpp"
+#include "interface/variable.hpp"
+#include "utils/utils.hpp"
+
+namespace parthenon {
+namespace impl {
+struct PackDescriptor {
+  std::vector<std::string> vars;
+  std::vector<bool> use_regex;
+  std::vector<MetadataFlag> flags;
+  bool with_fluxes;
+  bool coarse;
+
+  PackDescriptor(const std::vector<std::string> &vars, const std::vector<bool> &use_regex,
+                 const std::vector<MetadataFlag> &flags, bool with_fluxes, bool coarse)
+      : vars(vars), use_regex(use_regex), flags(flags), with_fluxes(with_fluxes),
+        coarse(coarse) {
+    PARTHENON_REQUIRE(use_regex.size() == vars.size(),
+                      "Must have a regex flag for each variable.");
+    PARTHENON_REQUIRE(!(with_fluxes && coarse),
+                      "Probably shouldn't be making a coarse pack with fine fluxes.");
+  }
+
+  PackDescriptor(const std::vector<std::pair<std::string, bool>> &vars_in,
+                 const std::vector<MetadataFlag> &flags, bool with_fluxes, bool coarse)
+      : flags(flags), with_fluxes(with_fluxes), coarse(coarse) {
+    for (auto var : vars_in) {
+      vars.push_back(var.first);
+      use_regex.push_back(var.second);
+    }
+    PARTHENON_REQUIRE(!(with_fluxes && coarse),
+                      "Probably shouldn't be making a coarse pack with fine fluxes.");
+  }
+
+  PackDescriptor(const std::vector<std::string> &vars_in,
+                 const std::vector<MetadataFlag> &flags, bool with_fluxes, bool coarse)
+      : vars(vars_in), use_regex(vars_in.size(), false), flags(flags),
+        with_fluxes(with_fluxes), coarse(coarse) {
+    PARTHENON_REQUIRE(!(with_fluxes && coarse),
+                      "Probably shouldn't be making a coarse pack with fine fluxes.");
+  }
+};
+} // namespace impl
+
+using namespace impl;
+
+class PackIdx {
+  std::size_t vidx;
+  int off;
```

Comment:
I like these at the bottom of the struct

Reply:
will move